solidity/contracts/bridge/VendingMachine.sol

```diff
@@ -0,0 +1,307 @@
+
+    /// @notice Divisor for precision purposes. Used to represent fractions
+    ///         in parameter values.
+    uint256 public constant FLOATING_POINT_DIVISOR = 1e18;
+
+    IERC20 public immutable tbtcV1;
+    TBTCToken public immutable tbtcV2;
+
+    /// @notice The fee for unminting TBTC v2 back into TBTC v1 represented as
+    ///         1e18 precision fraction. The fee is proportional to the amount
+    ///         being unminted and added on the top of the amount being unminted.
+    ///         To calculate the fee value, the amount being unminted needs
+    ///         to be multiplied by `unmintFee` and divided by 1e18.
+    ///         For example, `unmintFee` set to 1000000000000000
+    ///         means that 0.001 of the amount being unminted needs to be paid
+    ///         to the `VendingMachine` as an unminting fee on the top of the
+    ///         amount being unminted.
+    uint256 public unmintFee;
+    uint256 public newUnmintFee;
+    uint256 public unmintFeeUpdateInitiatedTimestamp;
+    address public unmintFeeUpdateInitiator;
+
+    /// @notice The address of a new vending machine. Set only when the update
+    ///         process is pending. Once the update gets finalized, the new
+    ///         vending machine will become an owner of TBTC v2 token.
+    address public newVendingMachine;
+    uint256 public vendingMachineUpdateInitiatedTimestamp;
+    address public vendingMachineUpdateInitiator;
+
+    event UnmintFeeUpdateInitiated(uint256 newUnmintFee, uint256 timestamp);
+    event UnmintFeeUpdated(uint256 newUnmintFee);
+
+    event VendingMachineUpdateInitiated(
+        address newVendingMachine,
+        uint256 timestamp
+    );
+    event VendingMachineUpdated(address newVendingMachine);
+
+    event Minted(address indexed recipient, uint256 amount);
+    event Unminted(address indexed recipient, uint256 amount, uint256 fee);
+
+    modifier only(address authorizedCaller) {
+        require(msg.sender == authorizedCaller, "Caller is not authorized");
+        _;
+    }
+
+    modifier onlyAfterGovernanceDelay(uint256 changeInitiatedTimestamp) {
+        GovernanceUtils.onlyAfterGovernanceDelay(
+            changeInitiatedTimestamp,
+            GOVERNANCE_DELAY
+        );
+        _;
+    }
+
+    constructor(
+        IERC20 _tbtcV1,
+        TBTCToken _tbtcV2,
+        uint256 _unmintFee
+    ) {
+        tbtcV1 = _tbtcV1;
+        tbtcV2 = _tbtcV2;
+        unmintFee = _unmintFee;
+
+        unmintFeeUpdateInitiator = msg.sender;
+        vendingMachineUpdateInitiator = msg.sender;
+    }
+
+    /// @notice Mints TBTC v2 to the caller from TBTC v1 with 1:1 ratio.
+    ///         The caller needs to have at least `amount` of TBTC v1 balance
+    ///         approved for transfer to the `VendingMachine` before calling
+    ///         this function.
+    /// @param amount The amount of TBTC v2 to mint from TBTC v1
+    function mint(uint256 amount) external {
```

Comment:
What do you think about renaming this function to e.g. `wrap` or `wrapTbtcV1`? We will want to upgrade the `VendingMachine` contract to let it mint tokens for new TBTC v2 deposits. In such case, the new `VendingMachine` contract will serve two purposes: (1) convert existing TBTCv1 tokens to TBTCv2 tokens, and  (2) mint new TBTCv2 tokens for new TBTC deposits. Sticking with the name initiated in this version of `VendingMachine` for case (1) may be confusing, as `mint` would best describe what case (2) does.

Reply:
I think it makes sense, especially that the user will *wrap* their v1 into v2 using v1 as collateral. Wrapping/unwrapping is like creating CDP which is a non-taxable event. I will revisit the wording.